### PR TITLE
Add `IfNotPresent` and `Ignore` SSA policies

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR extends the apply behaviour with two policies  `IfNotPresent` and `Ignore`.

Fix: #936 
Implementation and tests: https://github.com/fluxcd/pkg/pull/625

## Controlling the apply behavior of resources

To change the apply behaviour for specific Kubernetes resources, you can annotate them with:

| Annotation                          | Default    | Values                                                     | Role            |
|-------------------------------------|------------|------------------------------------------------------------|-----------------|
| `kustomize.toolkit.fluxcd.io/ssa`   | `Override` | - `Override`<br/>- `Merge`<br/>- `IfNotPresent`<br/>- `Ignore` | Apply policy    |
| `kustomize.toolkit.fluxcd.io/force` | `Disabled` | - `Enabled`<br/>- `Disabled`                                 | Recreate policy |
| `kustomize.toolkit.fluxcd.io/prune` | `Enabled`  | - `Enabled`<br/>- `Disabled`                                 | Delete policy   |

#### Override

The `Override` policy instructs the controller to reconcile the Kubernetes resources with the desired state (YAML manifests) defined in the Flux source (Git, OCI, Bucket).

#### Merge

The `Merge` policy instructs the controller to preserve the fields added by other tools to the Kubernetes resources managed by Flux.

The fields defined in the manifests applied by the controller will always be overridden, the `Merge` policy works only for adding new fields that don’t overlap with the desired state.

#### IfNotPresent

The `IfNotPresent` policy instructs the controller to only apply the Kubernetes resources if they are not present on the cluster.

This policy can be used for Kubernetes Secrets and ValidatingWebhookConfigurations managed by cert-manager, where Flux creates the resources with fields that are later on mutated by other controllers.

#### Ignore

The `Ignore` policy instructs the controller to skip applying Kubernetes resources even if they present in a Flux source (Git, OCI, Bucket).

### CODE REVIEW

File: docs/spec/v1/kustomizations.md [hunk 1]
```diff
@@ -1044,6 +1044,76 @@ cd apps/my-app
...
+The `Merge` policy instructs the controller to preserve the fields added by other tools to the
+Kubernetes resources managed by Flux.
+
+The fields defined in the manifests applied by the controller will always be overridden,
+the `Merge` policy works only for adding new fields that don’t overlap with the desired
+state.
+
+For lists fields which are atomic (e.g. `.spec.tolerations` in PodSpec), Kubernetes
+doesn't allow different managers for such fields, therefore any changes to these
+fields will be undone. For more context, please see the Kubernetes enhancement document:
```
Comment (hiddeco) [inline docs/spec/v1/kustomizations.md:1084]:
```suggestion
fields will be reverted. For more context, please see the Kubernetes enhancement document:
```

File: docs/spec/v1/kustomizations.md [hunk 2]
```diff
@@ -1044,6 +1044,76 @@ cd apps/my-app
...
+where Flux creates the resources with fields that are later on mutated by other controllers.
+
+##### Ignore
+
+The `Ignore` policy instructs the controller to skip applying Kubernetes resources
+even if they present in a Flux source (Git, OCI, Bucket).
+
+#### `kustomize.toolkit.fluxcd.io/force`
+
+When set to `Enabled` this policy instructs the controller to recreate the Kubernetes resources
```
Comment (hiddeco) [inline docs/spec/v1/kustomizations.md:1102]:
```suggestion
When set to `Enabled`, this policy instructs the controller to recreate the Kubernetes resources
```

File: docs/spec/v1/kustomizations.md [hunk 3]
```diff
@@ -1044,6 +1044,76 @@ cd apps/my-app
...
+with changes to immutable fields.
+
+This policy can be used for Kubernetes Jobs to rerun them when their container image changes.
+
+**Note** that using this policy for StatefulSets can incur data loss.
```
Comment (hiddeco) [inline docs/spec/v1/kustomizations.md:1107]:
```suggestion
**Note:** Using this policy for StatefulSets may result in potential data loss.
```

File: docs/spec/v1/kustomizations.md [hunk 4]
```diff
@@ -1044,6 +1044,76 @@ cd apps/my-app
...
+
+#### `kustomize.toolkit.fluxcd.io/prune`
+
+When set to `Disabled` this policy instructs the controller to skip the deletion of
```
Comment (hiddeco) [inline docs/spec/v1/kustomizations.md:1111]:
```suggestion
When set to `Disabled`, this policy instructs the controller to skip the deletion of
```

File: docs/spec/v1/kustomizations.md [hunk 5]
```diff
@@ -1044,6 +1044,76 @@ cd apps/my-app
...
 
+### Controlling the apply behavior of resources
+
+To change the apply behaviour for specific Kubernetes resources, you can annotate them with:
+
+| Annotation                          | Default    | Values                                                     | Role            |
+|-------------------------------------|------------|------------------------------------------------------------|-----------------|
+| `kustomize.toolkit.fluxcd.io/ssa`   | `Override` | -`Override`<br/>-`Merge`<br/>-`IfNotPresent`<br/>-`Ignore` | Apply policy    |
+| `kustomize.toolkit.fluxcd.io/force` | `Disabled` | -`Enabled`<br/>-`Disabled`                                 | Recreate policy |
+| `kustomize.toolkit.fluxcd.io/prune` | `Enabled`  | -`Enabled`<br/>-`Disabled`                                 | Delete policy   |
```
Comment (hiddeco) [inline docs/spec/v1/kustomizations.md:1055]:
```suggestion
| `kustomize.toolkit.fluxcd.io/ssa`   | `Override` | - `Override`<br/>- `Merge`<br/>- `IfNotPresent`<br/>- `Ignore` | Apply policy    |
| `kustomize.toolkit.fluxcd.io/force` | `Disabled` | - `Enabled`<br/>- `Disabled`                                 | Recreate policy |
| `kustomize.toolkit.fluxcd.io/prune` | `Enabled`  | - `Enabled`<br/>- `Disabled`                                 | Delete policy   |
```

**Review by hiddeco — APPROVED** [review]
Awesome to have the annotations documented in one place now :100: :1st_place_medal:

**Comment by hiddeco** [comment]
You may want to link from https://fluxcd.io/flux/faq/#how-to-patch-coredns-and-other-pre-installed-addons to the new documentation once this is available on the website.

**Comment by JeremyMordkoff** [comment]
I am struggling to get this to work for a configmap. 

We are using helm and it includes a configmap to which I have added this annotation. Then my master pod overwrites this configmap. This version also has the annotation. But flux still reconciles the configmap back the helm chart version. 

Does the annotation need to be in version 0 of this resource or can it be introduced in an upgrade? 
Could the issue be that the master pod is re-creating the configmap (as opposed to editing it ... I'm not sure about this part...just guessing that's what we do. 

I hope this is an appropriate place to ask ....

**Comment by stefanprodan** [comment]
SSA has nothing to do with Helm, these annotations are only for resources managed by Flux Kustomizations https://fluxcd.io/flux/components/kustomize/kustomizations/#controlling-the-apply-behavior-of-resources

**Pull request merged**